stablehlo/dialect/TypeInference.cpp

```diff
@@ -3366,6 +3368,80 @@ LogicalResult verifyCollectivePermuteOp(
   return success();
 }
 
+LogicalResult verifyCompositeOp(std::optional<Location> loc, Operation* op,
+                                StringRef name, StringRef decomposition) {
+  // composite_c1
+  llvm::SmallVector<llvm::StringRef> parts;
+  name.split(parts, '.');
+  if (
+      // Name must have at least 2 non-empty parts separated by periods.
+      parts.size() < 2 ||
+      llvm::any_of(parts, [](auto part) { return part.empty(); }) ||
+      // First part's first character must be alpha, and subsequent characters
+      // must be alpha, digit or undescore.
+      !isalpha(parts[0][0]) ||
+      !llvm::all_of(
+          llvm::drop_begin(parts[0]),
+          [](char c) { return isalpha(c) || isdigit(c) || c == '_'; }) ||
+      // Subsequent parts must be all alpha, digit, underscore or dollar sign.
+      !llvm::all_of(llvm::drop_begin(parts), [](llvm::StringRef part) {
+        return llvm::all_of(part, [](char c) {
+          return isalpha(c) || isdigit(c) || c == '_' || c == '$';
+        });
+      })) {
+    return emitOptionalError(loc,
+                             "name must be a valid namespaced op name, i.e. it",
+                             "must match the following regular expression: ",
+                             "[a-zA-Z][a-zA-Z0-9_]*([.][a-zA-Z0-9_$]+)+, e.g. "
+                             "\"my_namespace.my_op\"");
+  }
+
+  // composite_c2
+  auto module = op->getParentOfType<mlir::ModuleOp>();
+  auto decomp = module.lookupSymbol<mlir::func::FuncOp>(decomposition);
+  if (!decomp) {
+    return emitOptionalError(loc,
```

Comment:
edification: why is loc not from the op itself but comes in from the verify?

Reply:
It actually does come from the op: 

```
LogicalResult CompositeOp::verify() {
  return hlo::verifyCompositeOp(getLoc(), getOperation(), getName(),
                                getDecomposition());
}
```